### ISSUE TITLE
Infer array types when return type is given

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#663]
Similar to #660 could psalm also keep inferred types for arrays? Currently non-specific `array` return type hint seems to be overriding the inferred type:
```php
<?php
    // this complains
    /**
     * @psalm-return array<array{0:string,1:int}>
     */
    function f(): array
    {
        /** @var array{0:string,1:string}[] */
        $ret = [];
        return array_map(
          function (array $row): array {
            return [
                strval($row[0]),
                intval($row[1]),
            ];
        }, $ret);
    }

    // but this is fine
    /**
     * @psalm-return array<array{0:string,1:int}>
     */
    function f2()
    {
        /** @var array{0:string,1:string}[] */
        $ret = [];
        return array_map(
          function (array $row) {
            return [
                strval($row[0]),
                intval($row[1]),
            ];
        }, $ret);
    }
```
Ideally `array $row` parameter would also keep the type inferred in array_map call (at least when it's using the inline closure).
